Add "Schedule Part" enum sensor to Venstar thermostat

scolby33 opened the pull request:
## Proposed change

Add the `schedule_part` as an enum sensor to Venstar thermostat. This allows visibility into what schedule period is active, eliminating the need to hard code times in Home Assistant when wanting to trigger automations based on the thermostat's current schedule.


## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/25372

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.
  - https://github.com/home-assistant/core/pull/84310#pullrequestreview-1225415484
  - https://github.com/home-assistant/core/pull/80295#pullrequestreview-1225420509

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

scolby33 committed:
Add the `schedule_part` extra state attribute.

This allows visibility into what schedule period is active, eliminating
the need to hard code times in Home Assistant when wanting to trigger
automations based on the thermostat's current schedule.

scolby33 committed:
Update Venstar tests to include schedule_part attr.

scolby33 commented:
I'm running in to a dependency issue on macOS, `pyobjc-framework-cocoa 9.0.1 requires pyobjc-core>=9.0.1, but you'll have pyobjc-core 8.5.1 which is incompatible.`, so I can't run the tests locally. By inspection, I think this should be fine, but I'm looking forward to a maintainer approving this to run the workflows in CI.

cthoyt commented on homeassistant/components/venstar/const.py:
```diff
@@ -5,6 +5,7 @@
 
 ATTR_FAN_STATE = "fan_state"
 ATTR_HVAC_STATE = "hvac_mode"
+ATTR_SCHED_PART = "schedule_part"
```
why is there a mismatch between there being an underscore here and no underscore in `climate.py`?

scolby33 replied:
The Venstar API and, as a result, the venstarcolortouch library that Home Assistant is using, don't separate words in attribute names (similarly you can see `self._client.fanstate` in climate.py, where `_client` is an instance of a type from the venstarcolortouch library). However, the other attribute names here do separate words with underscores, so I followed that pattern.

I suspect that this is because the attribute names are user-facing in Home Assistant and this is the style they've adopted, whereas the underlying library is just copying the Venstar API's style, which is clearly not meant for the eyes of users.

cthoyt commented on homeassistant/components/venstar/climate.py:
```diff
@@ -192,6 +193,7 @@ def extra_state_attributes(self):
         return {
             ATTR_FAN_STATE: self._client.fanstate,
             ATTR_HVAC_STATE: self._client.state,
+            ATTR_SCHED_PART: self._client.get_info("schedulepart")
```
more specifically about the mismatch - why no underscore in `schedulepart` here?

garbled1 replied:
This looks correct to me actually, because that's how the API presents it.

cthoyt replied:
Yup, the comment in https://github.com/home-assistant/core/pull/84332#discussion_r1053957632 cleared things up!

scolby33 committed:
Add trailing comma for black.

garbled1 commented:
I'm good with this change. It looks to just be adding a new attribute, and even my ancient venstar has it, so I don't see where it would explode.

scolby33 commented:
@MartinHjelmare just bumping this to get approval for the CI to run on the newest commit

MartinHjelmare commented on homeassistant/components/venstar/climate.py:
```diff
@@ -192,6 +193,7 @@ def extra_state_attributes(self):
         return {
             ATTR_FAN_STATE: self._client.fanstate,
             ATTR_HVAC_STATE: self._client.state,
+            ATTR_SCHED_PART: self._client.get_info("schedulepart"),
```
What does the schedule part represent? Is it a time integer?

MartinHjelmare replied:
I checked the docs PR and it represents one of the available schedule times, which are parts of a say.

I think this should be an `enum` device class sensor instead.

https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

scolby33 replied:
I think you're correct on this choice. I have pushed changes to create an enum entity for the schedule part.

I will update the corresponding documentation PR once this is confirmed to be the correct path to take.

scolby33 committed:
Change Venstar schedule part from an attribute on the climate entity to a sensor in its own right.

scolby33 committed:
Don't try to async_add_entities() with an empty list.

scolby33 commented on homeassistant/components/venstar/sensor.py:
```diff
@@ -56,6 +56,14 @@
     RUNTIME_OV: "Override",
 }
 
+SCHEDULE_PARTS: dict[int, str] = {
```
Should this be a `defaultdict(lambda: "Unknown")` in order to defend against new values from the API?

MartinHjelmare replied:
I think it's better to just blow up so we know as soon as possible and can fix it.

scolby33 committed:
Fixes to Venstar schedule part enum sensor so it actually runs.

scolby33 committed:
Merge branch 'dev' into venstar_schedule_part_attr

MartinHjelmare commented on homeassistant/components/venstar/sensor.py:
```diff
@@ -76,31 +84,42 @@ async def async_setup_entry(
+    for description in INFO_ENTITIES:
+        try:
+            # just checking if the key exists
+            _ = coordinator.client.get_info(description.key)
```
```suggestion
            coordinator.client.get_info(description.key)
```

MartinHjelmare commented on homeassistant/components/venstar/sensor.py:
```diff
@@ -210,3 +229,16 @@ def native_unit_of_measurement(self) -> str | None:
+        key="schedulepart",
+        device_class=SensorDeviceClass.ENUM,
+        options=list(SCHEDULE_PARTS.values()),
+        uom_fn=lambda _: None,
```
This could be the default in the `VenstarSensorEntityDescription`.

scolby33 replied:
This would be neat, but the way the mixin is currently implemented means that there will be inter-mixed default and non-default arguments, which Python isn't happy with, so this would require more rearchitecting than I think is the goal here:
```diff
-     uom_fn: Callable[[Any], str | None]
+    uom_fn: Callable[[Any], str | None] = lambda _: None
```

`TypeError: non-default argument 'key' follows default argument`

MartinHjelmare replied:
We would move the attribute to the `VenstarSensorEntityDescription`.

This was a side note so we don't need to do it now.

MartinHjelmare commented on homeassistant/components/venstar/sensor.py:
```diff
@@ -210,3 +229,16 @@ def native_unit_of_measurement(self) -> str | None:
+        value_fn=lambda coordinator, sensor_name: SCHEDULE_PARTS[
+            coordinator.client.get_info(sensor_name)
+        ],
+        name_fn=lambda coordinator, sensor_name: f"{coordinator.client.name} Schedule Part",
```
Side note for the future: We recommend existing integrations to move to adhering to our entity naming guidelines and let the device info add the device name to the entity name by default.

https://developers.home-assistant.io/docs/core/entity#entity-naming

MartinHjelmare commented on homeassistant/components/venstar/sensor.py:
```diff
@@ -210,3 +229,16 @@ def native_unit_of_measurement(self) -> str | None:
+    VenstarSensorEntityDescription(
+        key="schedulepart",
+        device_class=SensorDeviceClass.ENUM,
+        options=list(SCHEDULE_PARTS.values()),
```
The options should be lowercase snake_case and have a translation via the `translation_key`. See eg:

https://github.com/home-assistant/core/blob/49b1d6e7fef780e77afe48cdc9b4c5155c9d7703/homeassistant/components/airvisual/sensor.py#L41-L55

https://github.com/home-assistant/core/blob/49b1d6e7fef780e77afe48cdc9b4c5155c9d7703/homeassistant/components/airvisual/strings.json#L67-L76

scolby33 committed:
Fixes based on PR review.

- Remove unnecessary assignment to `_`
- Use translation_key for names of schedule part states

MartinHjelmare reviewed (approved):
Thanks!
